lib/src/rules/require_trailing_comma.dart

```diff
@@ -0,0 +1,155 @@
+function literal implemented using curly braces, a literal map, a literal set or
+a literal array. This exception only applies if the final parameter does not fit
+entirely on one line.
+
+''';
+
+class RequireTrailingComma extends LintRule implements NodeLintRule {
+  RequireTrailingComma()
+      : super(
+          name: 'require_trailing_comma',
+          description: _desc,
+          details: _details,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(
+    NodeLintRegistry registry,
+    LinterContext context,
+  ) {
+    var visitor = _Visitor(this);
+    registry
+      ..addCompilationUnit(this, visitor)
+      ..addArgumentList(this, visitor)
+      ..addFormalParameterList(this, visitor)
+      ..addAssertStatement(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  static const _trailingCommaCode = LintCode(
+    'require_trailing_comma',
+    'Missing a required trailing comma.',
+  );
+
+  final LintRule rule;
+
+  LineInfo? _lineInfo;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) => _lineInfo = node.lineInfo;
+
+  @override
+  void visitArgumentList(ArgumentList node) {
+    super.visitArgumentList(node);
+    if (node.arguments.isEmpty) return;
+    _checkTrailingComma(
+      node.leftParenthesis,
+      node.rightParenthesis,
+      node.arguments.last,
+    );
+  }
+
+  @override
+  void visitFormalParameterList(FormalParameterList node) {
+    super.visitFormalParameterList(node);
+    if (node.parameters.isEmpty) return;
+    _checkTrailingComma(
+      node.leftParenthesis,
+      node.rightParenthesis,
+      node.parameters.last,
+    );
+  }
+
+  @override
+  void visitAssertStatement(AssertStatement node) {
+    super.visitAssertStatement(node);
+    _checkTrailingComma(
+      node.leftParenthesis,
+      node.rightParenthesis,
+      node.message ?? node.condition,
+    );
+  }
+
+  void _checkTrailingComma(
+    Token leftParenthesis,
+    Token rightParenthesis,
+    AstNode lastNode,
+  ) {
+    // Early exit if trailing comma is present.
+    if (lastNode.endToken.next?.type == TokenType.COMMA) return;
+
+    // No trailing comma is needed if the function call or definition, up to the
+    // closing parenthesis, fits on a single line. Ensuring the left and right
+    // parenthesis are on the same line is sufficient since dartfmt places the
+    // left parenthesis right after the identifier (on the same line).
+    if (_isSameLine(leftParenthesis, rightParenthesis)) return;
+
+    // Check the last parameter to determine if there are any exceptions.
+    if (_shouldAllowTrailingCommaException(lastNode)) return;
+
+    rule.reportLintForToken(rightParenthesis, errorCode: _trailingCommaCode);
+  }
+
+  bool _isSameLine(Token token1, Token token2) =>
+      _lineInfo!.getLocation(token1.charOffset).lineNumber ==
+      _lineInfo!.getLocation(token2.charOffset).lineNumber;
```

Comment:
Consider using `offset` rather than `charOffset`. It's slightly more efficient and more readable (imo).

Reply:
Changed `charOffset` to `offset`.